client/src/actions/bulk.js

```diff
@@ -67,10 +76,61 @@ function doAction(action, form, isAsyncJob) {
     };
 }
 
-export function bulkValidateMonuments(form) {
+export function bulkValidateSuggestions(form) {
     return doAction(actions.validate, form);
 }
 
+export function bulkCreateSuggestions(form) {
+    return doAction(actions.createSuggestion, form, true);
+}
+
 export function bulkCreateMonuments(form) {
-    return doAction(actions.create, form, true);
+    return async dispatch => {
+        dispatch(pending(actions.create));
+
+        try {
+            // Create the BulkCreateMonumentSuggestion
+            let suggestResult = await (await fetch(actions.create.suggestUri, {
+                method: 'post',
+                body: buildFormData(form)
+            })).json();
+
+            const suggestJobId = suggestResult.id;
+
+            let interval;
+            await new Promise(resolve => {
+                interval = window.setInterval(async () => {
+                    suggestResult = await (await fetch(`${actions.create.suggestUri}/progress/${suggestJobId}`)).json();
+                    // Divide the suggestion progress by 2 so it ranges from 0.0 to 0.50
+                    dispatch(pending(actions.create, (suggestResult.progress / 2)));
+
+                    if (suggestResult.future && suggestResult.future.done) resolve();
+                }, 200);
+            });
+            window.clearInterval(interval);
+
+            suggestResult = await (await fetch(`${actions.create.suggestUri}/result/${suggestJobId}`)).json();
+
+            // Approve the BulkCreateMonumentSuggestion
+            let approveResult = await put(`${actions.create.approveUri}/${suggestResult.id}/approve`);
+
+            const approveJobId = approveResult.id;
+
+            await new Promise(resolve => {
+                interval = window.setInterval(async () => {
+                    approveResult = await (await fetch(`${actions.create.approveUri}/approve/progress/${approveJobId}`)).json();
+                    // Divide the approve progress by 2 then add 0.50 so it ranges from 0.50 to 1.00
+                    dispatch(pending(actions.create, ((approveResult.progress / 2) + 0.50)));
+
+                    if (approveResult.future && approveResult.future.done) resolve();
+                }, 200);
+            });
+            window.clearInterval(interval);
+
+            approveResult = await (await fetch(`${actions.create.approveUri}/approve/result/${approveJobId}`)).json();
+            dispatch(success(actions.create, approveResult));
+        } catch (err) {
+            dispatch(error(actions.create, err));
+        }
+    }
```

Comment:
![image](https://user-images.githubusercontent.com/16584007/79133621-56904900-7d7a-11ea-93b4-21816f5d00b9.png)

Reply:
I guess what I meant is that there is logic in `doAction` to do singular AsyncJobs. Since `bulkCreateMonuments` does more than just that I suppose that wouldn't work. But are we even using that functionality of `doAction` anymore? You could remove it from `doAction` and instead extract some of the functionality in `bulkCreateMonuments` that's done twice (once for create, once for approve) into a utility function that `bulkCreateMonuments` calls twice.

You could also just delete the AsyncJob stuff from `doAction` and move on. I'm fine with that too.

Reply:
Yeah, it's still used in `doAction`. See line 83: `bulkCreateSuggestions`. Creating the Bulk-Create Suggestions and inserting them are both `AsyncJob`s. I'm happy to make the logic generic, but the logic calls `dispatch(pending(action, progress))`. Wouldn't I need to pass `dispatch` as a parameter?

Reply:
You can pass dispatch as a param, that's actually how redux-thunk reducers work (you pass a function that accepts the param `dispatch`, and thunk calls your function passing `dispatch` in). I don't think this refactoring is necessary